### PR TITLE
feat(tools): add OpenSandboxTool for safe code execution in CrewAI ag…

### PR DESCRIPTION
**Pull request by hellomypastor**
## Summary

- Add `OpenSandboxTool` — a new CrewAI tool that enables agents to execute Python code securely in isolated [[OpenSandbox](https://github.com/alibaba/OpenSandbox)](https://github.com/alibaba/OpenSandbox) containers
- Sandbox state (variables, files, installed packages) persists across multiple invocations within a single task, allowing agents to iteratively build up computation
- Supports on-demand library installation, configurable resource limits (CPU/memory), and custom sandbox images

## Motivation

CrewAI currently lacks a built-in sandboxed code execution capability. Agents that need to run code (data analysis, scripting, computation) either rely on unsafe `exec()` calls or require developers to implement their own isolation layer. `OpenSandboxTool` fills this gap by providing a production-ready, secure code execution tool out of the box.

## Changes

| File | Description |
|------|-------------|
| `open_sandbox_tool.py` | Core tool implementation — sandbox lifecycle management, async code execution, library installation |
| `__init__.py` | Export `OpenSandboxTool` in the public API |
| `pyproject.toml` | Add `opensandbox` optional dependency group |
| `test_open_sandbox_tool.py` | 23 unit tests covering execution, error handling, library install, sandbox reuse, and edge cases |

## Usage

```python
from crewai_tools import OpenSandboxTool

tool = OpenSandboxTool(
    opensandbox_domain="localhost:8080",
    opensandbox_image="opensandbox/code-interpreter:v1.0.2",
)

# Use in a CrewAI agent
agent = Agent(
    role="Data Analyst",
    tools=[tool],
    ...
)
```

## Test Plan

- [x] All 23 unit tests pass (`uv run pytest tests/tools/test_open_sandbox_tool.py`)
- [ ] Manual verification with a running OpenSandbox server
- [ ] Verify no regressions in existing tools (`uv run pytest`)

### CODE REVIEW

File: lib/crewai-tools/src/crewai_tools/tools/open_sandbox_tool/open_sandbox_tool.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+    _sandbox: Any = PrivateAttr(default=None)
+    _interpreter: Any = PrivateAttr(default=None)
+    _installed_libraries: set[str] = PrivateAttr(default_factory=set)
+
+    def _get_api_key(self) -> str | None:
+        return self.opensandbox_api_key or os.environ.get("OPENSANDBOX_API_KEY")
+
+    def _run(self, code: str, libraries: list[str] | None = None, **kwargs: Any) -> str:
+        try:
+            return asyncio.run(self._arun(code=code, libraries=libraries, **kwargs))
```
[the remark this answers is not in the document]
Comment (hellomypastor) [reply lib/crewai-tools/src/crewai_tools/tools/open_sandbox_tool/open_sandbox_tool.py]: > **The cross-loop issue is a valid catch — thanks!**
>
> Fixed in `e40f75fc`. The root cause was `asyncio.run()` spinning up a fresh event loop per `_run()` call while `_sandbox`/`_interpreter` stayed alive from a previous loop.
>
> The fix introduces a persistent event loop (`_get_loop()`) that all sync invocations share, so async SDK objects always run on the loop they were created on. For the nested-loop case (e.g. Jupyter), we dispatch to the persistent loop via a `ThreadPoolExecutor` worker thread to avoid `run_until_complete` inside a running loop.

File: lib/crewai-tools/src/crewai_tools/tools/open_sandbox_tool/open_sandbox_tool.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+        except RuntimeError:
+            # Already in an async event loop — use nest_asyncio or create new loop
+            loop = asyncio.new_event_loop()
+            try:
+                return loop.run_until_complete(
```
[the remark this answers is not in the document]
Comment (hellomypastor) [reply lib/crewai-tools/src/crewai_tools/tools/open_sandbox_tool/open_sandbox_tool.py]: > This was addressed together with the cross-loop fix in `e40f75fc`. The old fallback (`asyncio.new_event_loop()` + `run_until_complete` on the same thread) has been replaced: we now dispatch to a persistent loop via `ThreadPoolExecutor` when a loop is already running, which avoids the nested `run_until_complete` RuntimeError entirely.

**Comment by greysonlalonde** [comment]
Hey! This is much appreciated. We have new docs on tool publishing [here](https://docs.crewai.com/en/guides/tools/publish-custom-tools); please check those out and reach out if you have any questions.

**Comment by hellomypastor** [comment]
> Thanks @greysonlalonde for pointing this out! I've reviewed the [[tool publishing guide](https://docs.crewai.com/en/guides/tools/publish-custom-tools)](https://docs.crewai.com/en/guides/tools/publish-custom-tools) — makes sense to ship this as a standalone `crewai-opensandbox` package on PyPI rather than bundling it into the monorepo.
>
> I'll restructure accordingly and close this PR once the independent package is published. Appreciate the guidance!
